### PR TITLE
Shutdown hook should grab the lock if there is one

### PR DESCRIPTION
**Pull request by duke8253**
This was causing some crashes during shutdown.

### CODE REVIEW

**Comment by SolidWallOfCode** [comment]
Talk to @shinrich - she had the same problem and we tweaked a call (I think the mutex locking) to transparently deal with null mutexes.

Also, please work with @dyrock on plugin coordination - one of the points of that is to unify this kind of code so bugs like this don't have to stepped on one place at a time.

**Comment by shinrich** [comment]
Yes, we tidied up the mutex try lock to gracefully handle the null mutex case.  Need to remember where we made that change.  
Functionally, this fix does address the missing lock acquisition problem.

**Comment by SolidWallOfCode** [comment]
See #4926.

File: src/traffic_server/traffic_server.cc [hunk 1]
```diff
@@ -223,9 +223,19 @@ struct AutoStopCont : public Continuation {
   mainEvent(int /* event */, Event * /* e */)
   {
     APIHook *hook = lifecycle_hooks->get(TS_LIFECYCLE_SHUTDOWN_HOOK);
-    while (hook) {
-      hook->invoke(TS_EVENT_LIFECYCLE_SHUTDOWN, nullptr);
-      hook = hook->next();
+    if (hook != nullptr) {
+      if (hook->m_cont->mutex != nullptr) {
+        MUTEX_TAKE_LOCK(hook->m_cont->mutex, this_ethread());
```
Comment (ywkaras) [inline src/traffic_server/traffic_server.cc]: Comments in I_Lock.h indicate MUTEX_TAKE_LOCK is deprecated.  In any case, without a corresponding MUTEX_UNTAKE_LOCK invocation, the lock is never released.

File: src/traffic_server/traffic_server.cc [hunk 2]
```diff
@@ -223,9 +223,19 @@ struct AutoStopCont : public Continuation {
...
+        while (hook) {
+          hook->invoke(TS_EVENT_LIFECYCLE_SHUTDOWN, nullptr);
+          hook = hook->next();
+        }
+      } else {
+        while (hook) {
+          hook->invoke(TS_EVENT_LIFECYCLE_SHUTDOWN, nullptr);
+          hook = hook->next();
+        }
+      }
```
Comment (ywkaras) [inline src/traffic_server/traffic_server.cc]: I think what you meant is:
```
while (hook) {
  if (hook->m_cont->mutex) {
    SCOPED_MUTEX_LOCK(g, hook->m_cont->mutex, this_ethread());
    hook->invoke(TS_EVENT_LIFECYCLE_SHUTDOWN, nullptr);
  } else {
    hook->invoke(TS_EVENT_LIFECYCLE_SHUTDOWN, nullptr);
  }
  hook = hook->next();
}
```

Comment (duke8253) [reply src/traffic_server/traffic_server.cc]: yeah, also just remembered that hook changes.

File: src/traffic_server/traffic_server.cc [hunk 3]
```diff
@@ -224,7 +224,12 @@ struct AutoStopCont : public Continuation {
   {
     APIHook *hook = lifecycle_hooks->get(TS_LIFECYCLE_SHUTDOWN_HOOK);
     while (hook) {
-      hook->invoke(TS_EVENT_LIFECYCLE_SHUTDOWN, nullptr);
+      if (hook->m_cont->mutex) {
```
Comment (SolidWallOfCode) [inline src/traffic_server/traffic_server.cc]: Why not
```
SCOPED_MUTEX_LOCK(lock, hook->m_cont->mutex, this_ethread());
hook->invokle(TS_EVENT_LIFECYCLE_SHUTDOWN, nullptr);
```
?

Comment (ywkaras) [reply src/traffic_server/traffic_server.cc]: So the lock object will be a harmless do-nothing if the mutex is null?

Comment (SolidWallOfCode) [reply src/traffic_server/traffic_server.cc]: Yes. It was designed that way precisely to get this effect. Better to check the null in one place, and not every callsite.

Comment (ywkaras) [reply src/traffic_server/traffic_server.cc]: To be OCD about not holding mutexes longer than necessary, I would prefer:
```
while (hook) {
  {
    SCOPED_MUTEX_LOCK(lock, hook->m_cont->mutex, this_ethread());
    hook->invoke(TS_EVENT_LIFECYCLE_SHUTDOWN, nullptr);
  }
  hook = hook->next;
}
```

**Comment by duke8253** [comment]
[approve ci autest]

File: src/traffic_server/traffic_server.cc (hunk 3 above)
Comment (duke8253) [reply src/traffic_server/traffic_server.cc]: It's during shutdown, I think I want it to hold it for longer, might expose other problems in the code.

**Pull request merged**